src/zkevm_specs/state.py

```diff
@@ -0,0 +1,544 @@
+
+    # 3. value is a byte
+    assert_in_range(row.value, 0, 2**8 - 1)
+
+
+@is_circuit_code
+def check_stack(row: Row, row_prev: Row):
+    get_call_id = lambda row: row.keys[1]
+    get_stack_ptr = lambda row: row.keys[3]
+
+    # 0. Unused keys are 0
+    assert row.keys[2] == 0
+    assert row.keys[4] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # The first stack operation in a stack position is always a write (can't
+    # read if it isn't written before)
+    #
+    # When the set of all keys changes (first access of a stack position in a call)
+    # - It must be a WRITE
+    if not all_keys_eq(row, row_prev):
+        assert row.is_write == 1
+
+    # 2. stack_ptr in range
+    stack_ptr = get_stack_ptr(row)
+    assert_in_range(stack_ptr, 0, MAX_STACK_PTR)
+
+    # 3. stack_ptr only increases by 0 or 1
+    if row.tag() == row_prev.tag() and get_call_id(row) == get_call_id(row_prev):
+        stack_ptr_diff = get_stack_ptr(row) - get_stack_ptr(row_prev)
+        assert_in_range(stack_ptr_diff, 0, 1)
+
+
+@is_circuit_code
+def check_storage(row: Row, row_prev: Row):
+    get_addr = lambda row: row.keys[2]
+    get_storage_key = lambda row: row.keys[4]
+
+    # TODO: cold VS warm
+    # TODO: connection to MPT on first and last access for each (address, key)
+
+    # 0. Unused keys are 0
+    assert row.keys[1] == 0
+    assert row.keys[3] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # We add an extra write to set the value of the state in previous block, with rwc=0.
+    #
+    # When the set of all keys changes (first access of storage (address, key))
+    # - It must be a WRITE
+    if not all_keys_eq(row, row_prev):
+        assert row.is_write == 1 and row.rw_counter == 0
+
+
+@is_circuit_code
+def check_call_context(row: Row, row_prev: Row):
+    get_call_id = lambda row: row.keys[1]
+    get_field_tag = lambda row: row.keys[3]
+
+    # 0. Unused keys are 0
+    assert row.keys[2] == 0
+    assert row.keys[4] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_account(row: Row, row_prev: Row):
+    get_addr = lambda row: row.keys[2]
+    get_field_tag = lambda row: row.keys[3]
+
+    # 0. Unused keys are 0
+    assert row.keys[1] == 0
+    assert row.keys[4] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # We add an extra write to setup the value of the previous block, with rwc=0.
+    #
+    # When the set of all keys changes (first access of storage (address, AccountFieldTag))
+    # - It must be a WRITE
+    if not all_keys_eq(row, row_prev):
+        assert row.is_write == 1 and row.rw_counter == 0
+
+    # 2. nonce
+    if get_field_tag(row) == AccountFieldTag.Nonce:
+        get_nonce = lambda row: row.value
+        # When all the keys are equal in two consecutive a rows:
+        # - if WRITE, Nonce increases by 1
+        if all_keys_eq(row, row_prev) and row.is_write == 1:
+            assert get_nonce(row) - get_nonce(row_prev) == 1
```

Comment:
Maybe not, because in EVM circuit the constraint already does that explicitly, like:

https://github.com/appliedzkp/zkevm-specs/blob/d7bfc7ccfea8ffd2c38f7ebb5fe655cf84f9eda1/src/zkevm_specs/evm/execution/begin_tx.py#L33

Reply:
Ah, in that case this constraint can be just removed then?

Reply:
Yeah, and I guess for balance and storage we don't have special constraint, so perhaps all of them could be omitted.